src/enzyme_ad/jax/Passes/SimplifyAffineExprs.cpp

```diff
@@ -0,0 +1,480 @@
+              space, eqMat, ineqMat, isl_dim_set, isl_dim_div, isl_dim_param,
+              isl_dim_cst)),
+          cst};
+}
+
+using PosMapTy = llvm::MapVector<unsigned, unsigned>;
+
+struct AffineExprToIslAffConverter {
+  PosMapTy dimPosMap;
+  PosMapTy symPosMap;
+  isl_local_space *ls;
+  isl_ctx *ctx;
+
+  isl_aff *getIslAff(AffineExpr expr) {
+    if (auto bo = dyn_cast<AffineBinaryOpExpr>(expr)) {
+      isl_aff *lhs = getIslAff(bo.getLHS());
+      isl_aff *rhs = getIslAff(bo.getRHS());
+      switch (bo.getKind()) {
+      case mlir::AffineExprKind::Add:
+        return isl_aff_add(lhs, rhs);
+      case mlir::AffineExprKind::CeilDiv:
+        return isl_aff_ceil(isl_aff_div(lhs, rhs));
+      case mlir::AffineExprKind::FloorDiv:
+        return isl_aff_floor(isl_aff_div(lhs, rhs));
+      case mlir::AffineExprKind::Mod:
+        return isl_aff_mod_val(lhs, isl_aff_get_constant_val(rhs));
+      case mlir::AffineExprKind::Mul:
+        return isl_aff_mul(lhs, rhs);
+      default:
+        LLVM_DEBUG(llvm::dbgs()
+                   << "Unhandled kind " << (unsigned)bo.getKind() << "\n");
+        return nullptr;
+      }
+    } else if (auto c = dyn_cast<AffineConstantExpr>(expr)) {
+      return isl_aff_val_on_domain(isl_local_space_copy(ls),
+                                   isl_val_int_from_si(ctx, c.getValue()));
+    } else if (auto dim = dyn_cast<AffineDimExpr>(expr)) {
+      unsigned pos = dimPosMap[dim.getPosition()];
+      return isl_aff_var_on_domain(isl_local_space_copy(ls), isl_dim_set, pos);
+    } else if (auto sym = dyn_cast<AffineSymbolExpr>(expr)) {
+      unsigned pos = symPosMap[sym.getPosition()];
+      return isl_aff_var_on_domain(isl_local_space_copy(ls), isl_dim_param,
+                                   pos);
+    }
+    LLVM_DEBUG(llvm::dbgs() << "Unhandled expr " << expr << "\n");
+    return nullptr;
+  }
+};
+
+struct IslToAffineExprConverter {
+  MLIRContext *mlirContext;
+  unsigned symOffset;
+  PosMapTy dimPosMap;
+  PosMapTy symPosMap;
+
+  AffineExpr createOpBin(__isl_take isl_ast_expr *Expr) {
+    AffineExpr LHS, RHS, Res;
+    isl_ast_op_type OpType;
+    assert(isl_ast_expr_get_type(Expr) == isl_ast_expr_op &&
+           "isl ast expression not of type isl_ast_op");
+    assert(isl_ast_expr_get_op_n_arg(Expr) == 2 &&
+           "not a binary isl ast expression");
+
+    OpType = isl_ast_expr_get_op_type(Expr);
+
+    LHS = create(isl_ast_expr_get_op_arg(Expr, 0));
+    RHS = create(isl_ast_expr_get_op_arg(Expr, 1));
+
+    if (!LHS || !RHS) {
+      isl_ast_expr_free(Expr);
+      return nullptr;
+    }
+
+    switch (OpType) {
+    default:
+      llvm_unreachable("This is no binary isl ast expression");
+    case isl_ast_op_add:
+      Res = LHS + RHS;
+      break;
+    case isl_ast_op_sub:
+      Res = LHS - RHS;
+      break;
+    case isl_ast_op_mul:
+      Res = (LHS * RHS);
+      break;
+    case isl_ast_op_div:
```

Comment:
@ftynse Do you have any insight on the AffineExpr floordiv/ceildiv/mod semantics and how they match the isl_ast_expr semantics regarding whether they floor to -inf or 0 or other edge cases I have not thought of?